In [98]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [99]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [100]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [101]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [102]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h", "price_change_percentage_7d", 
            "price_change_percentage_14d", "price_change_percentage_30d", 
            "price_change_percentage_60d", "price_change_percentage_200d", 
            "price_change_percentage_1y"]])


In [103]:
# Create a DataFrame with the scaled data
df_market_scaled = pd.DataFrame( 
    market_data_scaled, 
    columns=[ "price_change_percentage_24h", "price_change_percentage_7d", 
            "price_change_percentage_14d", "price_change_percentage_30d", 
            "price_change_percentage_60d", "price_change_percentage_200d", 
            "price_change_percentage_1y"]) 

# Copy the crypto names from the original data
df_market_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index

df_market_scaled=df_market_scaled.set_index("coin_id")


# Display sample data
df_market_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [104]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [105]:
# Create an empty list to store the inertia values
inertia = []
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_scaled)
    inertia.append(k_model.inertia_)

In [106]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}
df_elbow = pd.DataFrame(elbow_data)

# Create a DataFrame with the data to plot the Elbow curve
df_elbow.head()


k     inertia
0  1  287.000000
1  2  198.571818
2  3  145.635748
3  4   79.022435
4  5   66.217780

In [134]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
scaled_plot = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)
scaled_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [109]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [110]:
# Fit the K-Means model using the scaled data
model.fit(df_market_scaled)

KMeans(n_clusters=4)

In [111]:
# Predict the clusters to group the cryptocurrencies using the scaled data
market_clusters = model.predict(df_market_scaled)

# Print the resulting array of cluster values.
print(market_clusters)

[0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 1 0 1 0 0 2 0 1 1 3
 1 0 0 1]


In [112]:
# Create a copy of the DataFrame
df_market_copy = df_market_data.copy()


In [113]:
# Add a new column to the DataFrame with the predicted clusters
df_market_scaled["market_clusters"] = market_clusters

# Display sample data
df_market_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  market_clusters  
coin_id                                                    
bitcoin                        -0.251637                0  
ethereum                       -0.199352                0  
tether                         -0.282061                0  
ripple                         -0.295546                0  
bitcoin-cash                   -0.270317                0

In [115]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
scaled_scatter = df_market_scaled.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="market_clusters",
    hover_cols = ["coin_id"], 
    title = "Scatter Plot Cryptocurrency "
)
scaled_scatter

:NdOverlay   [market_clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [116]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)


In [117]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_pca_data = pca.fit_transform(df_market_scaled)
# View the first five rows of the DataFrame. 
market_pca_data[:5]

array([[-0.67763569,  0.86298169,  0.46079754],
       [-0.58025234,  0.47198689,  0.9459328 ],
       [-0.50399638, -0.1518506 , -0.58731781],
       [-0.55344602, -0.20511743, -0.43065119],
       [-1.22973016,  2.08114658,  1.74959373]])

In [143]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
explained_variance_ratios = pca.explained_variance_ratio_

# Calculate the total explained variance
total_explained_variance = sum(explained_variance_ratios)
print(total_explained_variance)


0.8857606675202614


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:* Total Explained Variance of the three principal components:e 88.6%
PC1: 0.3847
PC2: 0.3271
PC3: 0.1739
Total Explained Variance=0.3847+0.3271+0.1739=0.8858.

In [119]:
# Create a new DataFrame with the PCA data.
df_market_pca = pd.DataFrame(market_pca_data, columns=["PC1", "PC2", "PC3"])
# Creating a DataFrame with the PCA data

# Copy the crypto names from the original data
df_market_pca["coin_id"] = df_market_data.index


# Set the coinid column as index
df_market_pca = df_market_pca.set_index("coin_id")

# Display sample data
df_market_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.677636  0.862982  0.460798
ethereum     -0.580252  0.471987  0.945933
tether       -0.503996 -0.151851 -0.587318
ripple       -0.553446 -0.205117 -0.430651
bitcoin-cash -1.229730  2.081147  1.749594

---

### Find the Best Value for k Using the PCA Data

In [120]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [121]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_pca)
    inertia_pca.append(k_model.inertia_)


In [122]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k, "inertia": inertia_pca}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)

In [147]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
plot_pca = df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve with PCA ", 
    xticks=k
)
plot_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:* 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:**  Based on the previous and current Elbow Curve, it looks like k=4 is still the correct one.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [128]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters= 4 , random_state=0)

In [129]:
# Fit the K-Means model using the PCA data
model.fit(df_market_pca)

KMeans(n_clusters=4, random_state=0)

In [130]:
# Predict the clusters to group the cryptocurrencies using the PCA data
market_clusters = model.predict(df_market_pca)
# Print the resulting array of cluster values.
print(market_clusters)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [131]:
# Create a copy of the DataFrame with the PCA data
market_pca_predictions_df = df_market_pca.copy()
# Add a new column to the DataFrame with the predicted clusters
market_pca_predictions_df["MarketCluster"]=market_clusters

# Display sample data
market_pca_predictions_df.head()

PC1       PC2       PC3  MarketCluster
coin_id                                                  
bitcoin      -0.677636  0.862982  0.460798              2
ethereum     -0.580252  0.471987  0.945933              2
tether       -0.503996 -0.151851 -0.587318              0
ripple       -0.553446 -0.205117 -0.430651              0
bitcoin-cash -1.229730  2.081147  1.749594              2

In [146]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_scatter = market_pca_predictions_df.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="MarketCluster",
   hover_cols = ["coin_id"],
    title = "Scatter plot with PCA"
)
pca_scatter

:NdOverlay   [MarketCluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [148]:
# Composite plot to contrast the Elbow curves
composite_elbow_plot = plot_pca + scaled_plot

# Display the composite plot
composite_elbow_plot

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [150]:
# Composite plot to contrast the clusters
composite_cluster_plot = scaled_scatter + pca_scatter

# Display the composite plot
composite_cluster_plot

:Layout
   .NdOverlay.I  :NdOverlay   [market_clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [MarketCluster]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** It simplifies the model making it easier to interpret cluster boundaries and relationships. But it's more noticeable in the scaled scatter plot that there is a large standard deviation from the clusters. When the data points are that far away from one another within a cluster it becomes less accurate.